### PR TITLE
Periodically remove expired auth tokens

### PR DESCRIPTION
**Pull request by chdorner**
Similar to the worker that removes expired auth tickets, this adds a new
worker which runs every hour and deletes expired auth tokens from the
database.

### CODE REVIEW

**Comment by chdorner** [comment]
[Periodically clear out expired access tokens](https://trello.com/c/QiTyqaP7/8-periodically-clear-out-expired-access-tokens)

File: tests/h/auth/worker_test.py [hunk 1]
```diff
@@ -34,8 +34,34 @@ def test_it_leaves_valid_tickets(self, db_session, factories):
...
-    def celery(self, patch, db_session):
-        cel = patch('h.auth.worker.celery', autospec=False)
-        cel.request.db = db_session
-        return cel
+
+@pytest.mark.usefixtures('celery')
+class TestDeleteExpiredTokens(object):
+    def test_it_removes_expired_tokens(self, db_session, factories):
+        tokens = [
+            Token('luke', expires=datetime(2014, 5, 6, 7, 8, 9)),
```
Comment (nickstenning) [inline tests/h/auth/worker_test.py]: You've got the `factories` fixture in the params here, but aren't using it. Probably best to add a factory for tokens and use it here?

File: tests/h/models/token_test.py [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+
+    def test_init_sets_expires(self):
+        expires = datetime.now()
+        assert Token('luke', expires=expires).expires == expires
+
+    def test_init_sets_expires_to_None_when_missing(self):
+        assert Token('luke').expires is None
+
+    def test_init_generates_value(self):
+        assert Token('lukek').value is not None
```
Comment (nickstenning) [inline tests/h/models/token_test.py]: `assert Token('lukek').value` would cover both these cases.

File: h/celery.py [hunk 3]
```diff
@@ -37,7 +37,11 @@
         'delete-expired-authtickets': {
             'task': 'h.auth.worker.delete_expired_auth_tickets',
             'schedule': timedelta(hours=1)
-        }
+        },
+        'delete-expired-authtokens': {
```
Comment (nickstenning) [inline h/celery.py]: Probably `delete-expired-tokens` for consistency with the model name.

File: h/models/token.py [hunk 4]
```diff
@@ -39,8 +39,9 @@ class Token(Base, mixins.Timestamps):
     #: A NULL value in this column indicates a token that does not expire.
     expires = sqlalchemy.Column(sqlalchemy.DateTime, nullable=True)
 
-    def __init__(self, userid):
+    def __init__(self, userid, expires=None):
```
Comment (nickstenning) [inline h/models/token.py]: Hmm. This is fine, but we should probably fix this properly -- i.e. not override the method signature of the default sqlalchemy constructor, and update uses of `Token(userid)` to do `Token(userid=userid)` instead.

Comment (chdorner) [inline h/models/token.py]: Sure, but that can happen in a separate PR as this one isn't introducing the override of the constructor.

Comment (nickstenning) [inline h/models/token.py]: Absolutely it can (I said as much in the review summary) 👍

Comment (chdorner) [inline h/models/token.py]: Had to do it anyway to not have to faff around with the new factory.

File: tests/common/factories.py [hunk 5]
```diff
@@ -115,3 +115,12 @@ class Meta:  # pylint: disable=no-init, old-style-class
...
         return self.user.userid
+
+
+class Token(ModelFactory):
+
+    class Meta:  # pylint: disable=no-init, old-style-class
+        model = models.Token
+        force_flush = True
+
+    userid = factory.LazyAttribute(lambda _: ('acct' + FAKER.user_name() + '@example.com'))
```
Comment (nickstenning) [inline tests/common/factories.py]: Missing colon after `acct` here.

File: tests/h/auth/worker_test.py [hunk 6]
```diff
@@ -34,8 +34,28 @@ def test_it_leaves_valid_tickets(self, db_session, factories):
...
+        worker.delete_expired_tokens()
+        assert db_session.query(Token).count() == 0
+
+    def test_it_leaves_valid_tickets(self, db_session, factories):
+        factories.Token(expires=datetime(2014, 5, 6, 7, 8, 9))
+        factories.Token(expires=(datetime.utcnow() + timedelta(hours=1)))
+
+        assert db_session.query(Token).count() == 2
+        worker.delete_expired_tokens()
+        assert db_session.query(Token).count() == 1
```
Comment (nickstenning) [inline tests/h/auth/worker_test.py]: Could we add a test that a token with `expires=None` isn't deleted, because those are developer tokens and it would be a bit embarrassing if we blew them all away.

File: tests/common/factories.py (hunk 5 above)
Comment (chdorner) [inline tests/common/factories.py]: Argh, sorry, should have spotted this. Fixed.

File: tests/h/auth/worker_test.py (hunk 6 above)
Comment (chdorner) [inline tests/h/auth/worker_test.py]: That is a brilliant idea! Added :)
